### ISSUE TITLE
High resource usage on 2.46.0 with native-histograms enabled

### ISSUE THREAD

**Issue by benclapp** [prometheus/prometheus#12603]
### What did you do?

Upgraded from v2.45.0 to v2.46.0 was the only change initially. Running with the latest docker image. 

### What did you expect to see?

The same, if not improved resource usage after the upgrade

### What did you see instead? Under which circumstances?

Immediately after the upgrade, increased CPU and Memory usage. Query engine timings were also slower. Both memory and queries increased over time too. 

Disabling the `native-histograms` feature flag, at the point highlighted by cursor in the below screenshot, returned everything to expected behaviour. 

Here's a chart for reference. ![Chart for reference](https://cdn.zappy.app/d697c886a4f3fc98a3326bd487d1c12a.png)

### System information

Linux 5.4.0-155-generic x86_64

### Prometheus version

```text
prometheus, version 2.46.0 (branch: HEAD, revision: cbb69e51423565ec40f46e74f4ff2dbb3b7fb4f0)
  build user:       root@42454fc0f41e
  build date:       20230725-12:31:24
  go version:       go1.20.6
  platform:         linux/amd64
  tags:             netgo,builtinassets,stringlabels
```

First few log lines too, containing similar info

```
ts=2023-07-25T22:18:15.689Z caller=main.go:585 level=info msg="Starting Prometheus Server" mode=server version="(version=2.46.0, branch=HEAD, revision=cbb69e51423565ec40f46e74f4ff2dbb3b7fb4f0)"
ts=2023-07-25T22:18:15.693Z caller=main.go:590 level=info build_context="(go=go1.20.6, platform=linux/amd64, user=root@42454fc0f41e, date=20230725-12:31:24, tags=netgo,builtinassets,stringlabels)"
ts=2023-07-25T22:18:15.694Z caller=main.go:591 level=info host_details="(Linux 5.4.0-155-generic #172-Ubuntu SMP Fri Jul 7 16:10:02 UTC 2023 x86_64 bcb52a8c481e (none))"
ts=2023-07-25T22:18:15.694Z caller=main.go:592 level=info fd_limits="(soft=1048576, hard=1048576)"
ts=2023-07-25T22:18:15.698Z caller=main.go:593 level=info vm_limits="(soft=unlimited, hard=unlimited)"
```

### Prometheus configuration file

I don't believe config file is relevant but can provide if requested. Flags passed to Prometheus below. Removing `--enable-feature=native-histograms` fixed the issue

```yaml
command:
  - "--config.file=/etc/prometheus/prometheus.yml"
  - "--storage.tsdb.path=/prometheus"
  - "--storage.tsdb.retention.time=3d"
  - "--storage.tsdb.retention.size=75GB"
  - "--storage.tsdb.min-block-duration=2h"
  - "--storage.tsdb.max-block-duration=2h"
  - "--web.enable-lifecycle"
  - "--web.enable-admin-api"
  - "--web.page-title=REDACTED"
  - "--web.external-url=REDACTED"
  - "--enable-feature=native-histograms"
```

### Alertmanager version

```text

```

### Alertmanager configuration file

```yaml

```

### Logs

Potentially related - logs from startup after disabling native-histogram feature flag. 

```text
ts=2023-07-25T22:18:16.560Z caller=head.go:595 level=info component=tsdb msg="Replaying on-disk memory mappable chunks if any"
ts=2023-07-25T22:18:17.460Z caller=head.go:661 level=error component=tsdb msg="Loading on-disk chunks failed" err="iterate on on-disk chunks: out of sequence m-mapped chunk for series ref 18120713, last chunk: [1690321802579, 1690322397587], new: [1690321802579, 1690322397587]"
ts=2023-07-25T22:18:17.461Z caller=head.go:901 level=info component=tsdb msg="Deleting mmapped chunk files"
ts=2023-07-25T22:18:17.479Z caller=head.go:907 level=info component=tsdb msg="Deleting mmapped chunk files"
ts=2023-07-25T22:18:17.480Z caller=head.go:910 level=info component=tsdb msg="Deletion of corrupted mmap chunk files failed, discarding chunk files completely" err="cannot handle error: iterate on on-disk chunks: out of sequence m-mapped chunk for series ref 18120713, last chunk: [1690321802579, 1690322397587], new: [1690321802579, 1690322397587]"
ts=2023-07-25T22:18:17.564Z caller=head.go:676 level=info component=tsdb msg="On-disk memory mappable chunks replay completed" duration=1.003443018s
```

**Comment by beorn7** [comment]
As long as you don't actually ingest any native histograms, the `--enable-feature=native-histograms` doesn't really change anything. I would strongly assume that you have targets that already expose native histograms. That would also explain why you see fewer samples ingested (all the buckets and count and sum of a classic histogram are bundled into one sample with the native histograms).

And maybe those targets, that also expose native histograms, are using a very high resolution, and maybe you are running queries (in a recording rule?) that touch those high-res native histograms.

All speculation of course. First step would be to look at your targets and what they expose.

**Comment by benclapp** [comment]
Thanks for the suggestions! I've tried a few of these and I am still somewhat stuck. I have narrowed down a few things, though. 

- I don't think any of the targets are exposing native histograms, as far as I'm aware. This query returns `0` for `type="histogram"` `max_over_time(prometheus_tsdb_head_samples_appended_total{job="prometheus"}[1w])`. The first panel in the screenshot above is also broken out by the `type` label, too.
- Disabled all recording rules to verify they're not the cause, and that had no impact on resource usage. With no queries at all 
- The cause for the drop in samples ingested, was 4 targets that were in an `unknown` status and not being scraped at all it seems. They were the latest releases of Grafana, Promtail, and Thanos Query and Thanos Query Frontend (other Thanos components were fine, though). I recognise these aren't the responsibility of the Prometheus project. But for the record they're using `client_golang` versions v1.14.0, v1.15.1, and v1.16.0.
  - Had a quick scan of these repos, it doesn't _appear_ they've configured any histograms to switch from classic to native. A package they use could of course have native histograms however.

I am able to reliably reproduce with the following docker-compose and prometheus configs. Everything runs fine at startup. Once there's one request to one of the Thanos containers, their scrapes stop. 

From a before/after diff of the metrics exposed, looks like there are a few HTTP request/response duration/size Histograms exposed after the first request. 

Using a 2.45.0 version of Prometheus with native histograms enabled, doesn't seem to see any native histograms.

Other symptoms, POST requests to `/-/reload` hang.

<details>
  <summary>docker-compose.yaml</summary>

```yaml
version: "3"
services:
  prometheus:
    image: quay.io/prometheus/prometheus:v2.46.0
    container_name: prometheus
    restart: always
    volumes:
      - ./prometheus.yaml:/etc/prometheus/prometheus.yaml:ro
    command:
      - "--config.file=/etc/prometheus/prometheus.yaml"
      - "--storage.tsdb.retention.time=3d"
      - "--storage.tsdb.retention.size=75GB"
      - "--storage.tsdb.min-block-duration=2h"
      - "--storage.tsdb.max-block-duration=2h"
      - "--web.enable-lifecycle"
      - "--web.enable-admin-api"
      - "--enable-feature=native-histograms"
    ports:
      - 9090:9090

  grafana:
    container_name: grafana
    image: grafana/grafana:10.0.3
    restart: always
    ports:
      - 3000:3000

  thanos-query-frontend:
    image: quay.io/thanos/thanos:v0.31.0
    container_name: thanos-query-frontend
    restart: always
    ports:
      - 3001:3001
    command:
      - query-frontend
      - --http-address=0.0.0.0:3001
      - --query-frontend.downstream-url=http://thanos-query:3002
      - |
        --query-range.response-cache-config=type: IN-MEMORY
        config:
          max_size: 200MB
          validity: 24h

  thanos-query:
    image: quay.io/thanos/thanos:v0.31.0
    container_name: thanos-query
    restart: always
    ports:
      - 3002:3002 # http
      - 3003:3003 # cluster address
    command:
      - query
      - --http-address=0.0.0.0:3002
      - --grpc-address=0.0.0.0:3003
      - --query.default-evaluation-interval=5s
      - --query.replica-label=replica

```

</details>

<details>
  <summary>prometheus.yaml</summary>

```yaml
global:
  scrape_interval: 5s

scrape_configs:
- job_name: prometheus
  static_configs:
  - targets: ['localhost:9090']

- job_name: test
  static_configs:
  - targets:
    - grafana:3000
    - thanos-query-frontend:3001
    - thanos-query:3002

```

</details>

This is what I've discovered for now, will keep looking into it tomorrow perhaps with a bit more time.

**Comment by beorn7** [comment]
Thanks for reporting. Please continue your investigation. Maybe CPU and memory profiles and looking at the differences between the feature flag enabled and disabled will help.

As said, the feature flag really only gates the decision if native histograms exposed by a target are parsed as such or ignored. Therefore, any other change in behavior is very surprising, especially if your targets don't even expose native histograms. Something weird must be happening here.

**Comment by krajorama** [comment]
I've ran into this myself and managed to grab some heap profiles which point to AppendExemplar function in tsdb head. Which was recently modified by https://github.com/prometheus/prometheus/pull/12557 
![image](https://github.com/prometheus/prometheus/assets/13435893/5b11c38c-ed01-4faf-b334-d864434a3bda)

**Comment by krajorama** [comment]
Well, AppendExemplar was not modified but the code calling it now can run in a loop:
https://github.com/prometheus/prometheus/blob/cbb69e51423565ec40f46e74f4ff2dbb3b7fb4f0/scrape/scrape.go#L1688

That loop is not a loop for the text format, due to Exemplar->bool being false always, only for the protobuf which we switch to with native histograms enabled...

**Comment by krajorama** [comment]
In my case I see this loop go into an infinite loop (or at least above 1 million). So maybe not the same issue as here, so a heap profile for the issue reported here would be welcome.

**Comment by krajorama** [comment]
Issue and PR opened for the infinite loop case: #12731

**Comment by benclapp** [comment]
Sorry for ghosting, this slipped my radar while I was away on holiday. It looks like my configuration was fixed in the 2.47.0 release as well though, looks like by https://github.com/prometheus/prometheus/pull/12737.

Thanks both for investigating and solving!
